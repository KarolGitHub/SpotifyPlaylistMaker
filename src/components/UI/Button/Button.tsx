import React from 'react';
import classes from './Button.module.scss';

type Props = {
  btnType?: string;
  table?: boolean;
  clicked?: (event: any) => void;
  disabled?: boolean;
  type?: 'button' | 'submit';
  children?: any;
};
const Button: React.FC<Props> = ({
  btnType,
  table,
  children,
  clicked,
  type,
  disabled,
}) => {
  let btnClassType;
  switch (btnType) {
    case 'blue':
      btnClassType = ` ${classes.Blue} `;
      break;
    case 'purple':
      btnClassType = ` ${classes.Purple} `;
      break;
    case 'red':
      btnClassType = ` ${classes.Red} `;
      break;
    case 'green':
      btnClassType = ` ${classes.Green} `;
      break;
    case 'cancel':
      btnClassType = ` ${classes.Cancel} `;
      break;
    default:
      btnClassType = ``;
  }
  const btnClasses = `${classes.Button}${btnClassType}${
    table ? ' ' + classes.TableButton : ''
  }`;

  return (
    <button
      className={btnClasses}
      onClick={clicked}
      disabled={disabled}
      type={type ? type : 'button'}
    >
      {children}
    </button>
  );
};

export default Button;
