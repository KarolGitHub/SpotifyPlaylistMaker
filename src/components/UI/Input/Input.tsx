import React, { FunctionComponent, useEffect, useRef } from 'react';
import classes from './Input.module.scss';
import { isMobile } from '../../../shared/utility';

type Props = {
  type?: 'input' | 'select';
  value: string;
  changed: (event: { target: any }) => void;
  pressed?: (event: any) => void;
  invalid?: boolean;
  elementConfig: any;
  focus: boolean;
  note?: string;
};
const Input: FunctionComponent<Props> = ({
  type,
  value,
  changed,
  pressed,
  invalid,
  elementConfig,
  focus,
  note,
}) => {
  let inputClasses = `${classes.InputElement}`;
  const inputRef = useRef<HTMLInputElement>();
  let validationError = null;
  let inputElement = null;

  if (invalid) {
    validationError = (
      <p className={classes.ValidationError}>Please enter a proper value</p>
    );
    inputClasses = `${classes.InputElement} ${classes.Invalid}`;
  }

  switch (type) {
    case 'input':
      inputElement = (
        <input
          className={inputClasses}
          value={value}
          onBlur={() => invalid}
          onChange={changed}
          ref={(el: any) => (inputRef.current = el)}
          {...elementConfig}
          title={note}
        />
      );
      break;
    case 'select':
      inputElement = (
        <select
          className={inputClasses}
          value={value}
          onBlur={() => invalid}
          onChange={changed}
          title={note}
        >
          {elementConfig.options.map((option: any) => (
            <option key={option.value} value={option.value}>
              {option.displayValue}
            </option>
          ))}
        </select>
      );
      break;
    default:
      inputElement = (
        <input
          value={value}
          onChange={changed}
          onKeyUp={pressed}
          onBlur={() => invalid}
          className={inputClasses}
          ref={(el: any) => (inputRef.current = el)}
          {...elementConfig}
        />
      );
  }

  useEffect(() => {
    if (!isMobile && focus) {
      inputRef.current?.focus();
    }
  }, [focus]);

  return (
    <div className={classes.Input}>
      {inputElement}
      {validationError}
    </div>
  );
};

export default Input;
