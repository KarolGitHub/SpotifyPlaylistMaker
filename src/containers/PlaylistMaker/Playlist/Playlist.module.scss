@import '../../../assets/variables';

.Playlist {
  display: flex;
  flex-direction: column;
  align-items: center;
  overflow-y: scroll;
  width: 40%;
  padding: 14px;
  background-color: var(--playlist);
  box-shadow: 0 4px 2px 2px $gray;
  .Save {
    margin-top: 20px;
  }
}

@media only screen and (max-width: 800px) {
  .Playlist {
    height: unset;
    min-height: max-content;
    width: 90%;
    margin-bottom: 2rem;
  }
}