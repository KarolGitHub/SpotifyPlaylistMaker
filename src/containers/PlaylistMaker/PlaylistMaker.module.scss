.PlaylistMaker {
  height: 100%;
  padding: 0 5% 5% 5%;

  .Playlist {
    display: flex;
    justify-content: space-between;
    width: 100%;
  }
}

@media only screen and (min-width: 1000px) {
  .PlaylistMaker {
    max-width: 1000px;
    margin: auto;
  }
}

@media only screen and (max-width: 800px) {
  .PlaylistMaker {
    max-width: 600px;
    margin: auto;
    .Playlist {
      align-items: center;
      flex-direction: column;
    }
  }
}