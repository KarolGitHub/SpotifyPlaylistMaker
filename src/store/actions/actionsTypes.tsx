export const ADD_TRACK = "ADD_TRACK";
export const DELETE_TRACK = "DELETE_TRACK";
export const SET_TRACKS = "SET_TRACKS";

export const SEARCH_TRACKS_START = "SEARCH_TRACKS_START";
export const SEARCH_TRACKS = "SEARCH_TRACKS";
export const SEARCH_TRACKS_SUCCESS = "SEARCH_TRACKS_SUCCESS";
export const SEARCH_TRACKS_FAIL = "SEARCH_TRACKS_FAIL";

export const SAVE_PLAYLIST_START = "SAVE_PLAYLIST_START";
export const SAVE_PLAYLIST = "SAVE_PLAYLIST";
export const SAVE_PLAYLIST_SUCCESS = "SAVE_PLAYLIST_SUCCESS";
export const SAVE_PLAYLIST_FAIL = "SAVE_PLAYLIST_FAIL";
export const SUCCESS_CONFIRM = "SUCCESS_CONFIRM";

export const EDIT_PLAYLIST_START = "EDIT_PLAYLIST_START";
export const EDIT_PLAYLIST = "EDIT_PLAYLIST";
export const EDIT_PLAYLIST_SUCCESS = "EDIT_PLAYLIST_SUCCESS";
export const EDIT_PLAYLIST_FAIL = "EDIT_PLAYLIST_FAIL";

export const AUTH_START = "AUTH_START";
export const AUTH_SUCCESS = "AUTH_SUCCESS";
export const AUTH_FAIL = "AUTH_FAIL";
export const AUTH_LOGOUT = "AUTH_LOGOUT";
export const SET_AUTH_REDIRECT_URL = "SET_AUTH_REDIRECT_URL";
export const AUTH_CHECK_STATE = "AUTH_CHECK_STATE";
export const AUTH_CHECK_TIMEOUT = "AUTH_CHECK_TIMEOUT";

export const PLAY_TRACK_STATE_UPDATE = "PLAY_TRACK_STATE_UPDATE";
export const PLAY_TRACK_START = "PLAY_TRACK_START";
export const PLAY_TRACK_PAUSE = "PLAY_TRACK_PAUSE";
export const PLAY_TRACK_END = "PLAY_TRACK_END";
export const PLAY_TRACK_FAIL = "PLAY_TRACK_FAIL";

export const FETCH_PLAYLISTS_START = "FETCH_PLAYLISTS_START";
export const FETCH_PLAYLISTS = "FETCH_PLAYLISTS";
export const FETCH_PLAYLISTS_SUCCESS = "FETCH_PLAYLISTS_SUCCESS";
export const FETCH_PLAYLISTS_FAIL = "FETCH_PLAYLISTS_FAIL";
export const CLEAR_PLAYLIST = "CLEAR_PLAYLIST";
export const SET_REDIRECT = "SET_REDIRECT";

export const FETCH_TRACKS_START = "FETCH_TRACKS_START";
export const FETCH_TRACKS = "FETCH_TRACKS";
export const FETCH_TRACKS_SUCCESS = "FETCH_TRACKS_SUCCESS";
export const FETCH_TRACKS_FAIL = "FETCH_TRACKS_FAIL";

export const DELETE_TRACKS_START = "DELETE_TRACKS_START";
export const DELETE_TRACKS = "DELETE_TRACKS";
export const DELETE_TRACKS_SUCCESS = "DELETE_TRACKS_SUCCESS";
export const DELETE_TRACKS_FAIL = "DELETE_TRACKS_FAIL";
