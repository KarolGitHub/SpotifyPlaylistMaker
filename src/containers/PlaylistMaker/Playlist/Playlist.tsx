import React, { FunctionComponent } from "react";

import classes from "./Playlist.module.scss";
import Button from "../../../components/UI/Button/Button";
import TrackList from "../../../components/Playlist/TrackList/Tracklist";
import { Tracklist } from "../../../shared/utility";

type Props = {
  tracklist: Tracklist;
  name: string | null;
  clicked: any;
};
const Playlist: FunctionComponent<Props> = ({ tracklist, name, clicked }) => {
  return (
    <div className={classes.Playlist}>
      <h1>{name ? name : "New Playlist"}</h1>
      <TrackList tracklist={tracklist} listType="playlist" />
      <div className={classes.Save}>
        <Button
          btnType="purple"
          clicked={clicked}
          disabled={tracklist.length < 1}
        >
          Save To Spotify
        </Button>
      </div>
    </div>
  );
};

export default Playlist;
