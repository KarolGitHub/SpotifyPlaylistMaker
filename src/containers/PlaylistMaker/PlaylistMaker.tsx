import React, {
  FunctionComponent,
  useState,
  useEffect,
  useCallback,
  useMemo,
  lazy,
} from "react";
import { useSelector, useDispatch } from "react-redux";
import { DragDropContext } from "react-beautiful-dnd";

import classes from "./PlaylistMaker.module.scss";
import SearchBar from "./SearchBar/SearchBar";
import Playlist from "./Playlist/Playlist";
import SearchResults from "./SearchResults/SearchResults";
import Spinner from "./../../components/UI/Spinner/Spinner";
import Modal from "./../../components/UI/Modal/Modal";
import Checkout from "../../components/Playlist/Checkout/Checkout";
import {
  Track,
  Tracklist,
  Tuple,
  authPopup,
  PlaylistPayload,
  PlaylistInfo,
  moveItems,
  reorderItems,
} from "./../../shared/utility";
import { RootState } from "../../index";
import errorHandler from "./../../Hoc/errorHandler/errorHandler";
import axios from "./../../axios-spotify";
import * as actions from "./../../store/actions/index";
import Button from "../../components/UI/Button/Button";

const Player = lazy(() => {
  return import("./Player/Player");
});

const PlaylistMaker: FunctionComponent = () => {
  const [isModal, setModal] = useState(false);
  const [redirect, setRedirect] = useState(false);

  const tracks: Tracklist = useSelector((state: RootState) => {
    return state.playlistMaker.playlist;
  });
  const results: Tracklist = useSelector((state: RootState) => {
    return state.playlistMaker.searchResults;
  });
  const searchResultsLimit: number = useSelector((state: RootState) => {
    return state.playlistMaker.searchResultsLimit;
  });
  const loading: boolean = useSelector((state: RootState) => {
    return state.playlistMaker.loading;
  });
  const error: any = useSelector((state: RootState) => {
    return state.playlistMaker.error;
  });
  const saveResult: boolean = useSelector((state: RootState) => {
    return state.playlistMaker.saveResult;
  });
  const token: string = useSelector((state: RootState) => {
    return state.auth.token;
  });
  const userId: string = useSelector((state: RootState) => {
    return state.auth.userId;
  });
  const country: string = useSelector((state: RootState) => {
    return state.auth.country;
  });
  const authRedirectPath: string = useSelector((state: RootState) => {
    return state.auth.authRedirectPath;
  });
  const playerState: Tuple = useSelector((state: RootState) => {
    return state.player.playerState;
  });
  const playlistInfo: PlaylistInfo | null = useSelector((state: RootState) => {
    return state.playlists.playlistInfo;
  });

  const dispatch = useDispatch();

  const onTracksSearch = useCallback(
    (type: string, term: string, limit: number) =>
      dispatch(actions.searchTracks(token, country, type, term, limit)),
    [dispatch, token, country]
  );
  const onPlaylistSave = useCallback(
    (payload: PlaylistPayload, trackURIs: Array<string>) =>
      dispatch(actions.savePlaylist(token, userId, payload, trackURIs)),
    [dispatch, token, userId]
  );
  const onSuccessConfirm = useCallback(
    () => dispatch(actions.successConfirm()),
    [dispatch]
  );
  const onClearPlaylist = useCallback(() => dispatch(actions.clearPlaylist()), [
    dispatch,
  ]);
  const onEditPlaylist = useCallback(
    (payload, trackURIs, changedDetails) =>
      dispatch(
        actions.editPlaylist(
          token,
          userId,
          payload,
          trackURIs,
          changedDetails,
          playlistInfo?.id
        )
      ),
    [dispatch, token, userId, playlistInfo]
  );
  const onSetTracks = useCallback(
    (tracklist: {}) => dispatch(actions.setTracks(tracklist)),
    [dispatch]
  );
  const onMoveTrack = useCallback(
    (tracklist: {}) => dispatch(actions.moveTrack(tracklist)),
    [dispatch]
  );

  const compareDetails = (
    userPayload: PlaylistPayload,
    playlistPayload: PlaylistPayload
  ) => [JSON.stringify(userPayload) !== JSON.stringify(playlistPayload)];

  const modalHandler = (prevState: boolean) => {
    setModal(!prevState);
  };

  const confirmModalHandler = useCallback(
    (prevState: boolean) => {
      onSuccessConfirm();
      setModal(!prevState);
      onClearPlaylist();
    },
    [onSuccessConfirm, onClearPlaylist]
  );

  const savePlaylistHandler = useCallback(
    (payload: PlaylistPayload) => {
      const trackURIs: Array<string> = [];
      tracks.map((track: Track) => trackURIs.push(track.uri));
      if (token) {
        if (playlistInfo) {
          const changedDetails = compareDetails(payload, playlistInfo.payload);
          onEditPlaylist(payload, trackURIs, changedDetails);
        } else {
          onPlaylistSave(payload, trackURIs);
        }
      } else {
        setRedirect(true);
      }
    },
    [token, tracks, onPlaylistSave, playlistInfo, onEditPlaylist]
  );
  const searchHandler = useCallback(
    (type: string, term: string, limit: number) => {
      if (token) {
        onTracksSearch(type, term, limit);
      } else {
        setRedirect(true);
      }
    },
    [onTracksSearch, token, setRedirect]
  );

  const dragEndHandler = (result: any) => {
    const { source, destination } = result;

    if (!destination) {
      return;
    }
    const sInd = source.droppableId;
    const dInd = destination.droppableId;
    const sourceItems = sInd === "playlist" ? tracks : results;
    if (sInd === dInd) {
      const updatedItems: {} = reorderItems(
        sourceItems,
        destination.index,
        source
      );
      onSetTracks(updatedItems);
    } else {
      const destinationItems = dInd === "playlist" ? tracks : results;
      const updatedItems: {} = moveItems(
        sourceItems,
        destinationItems,
        source,
        destination
      );
      onMoveTrack(updatedItems);
    }
  };

  useEffect(() => {
    if (redirect) {
      window.addEventListener("storage", () => {
        const storage = window.localStorage.getItem("token");
        if (storage !== null) {
          dispatch(actions.authCheckState());
          window.removeEventListener("storage", () => {});
        }
      });
      authPopup(authRedirectPath);
    }
  }, [dispatch, redirect, authRedirectPath]);
  useEffect(() => {
    if (!authRedirectPath) {
      dispatch(actions.setAuthRedirectURL());
    }
  }, [authRedirectPath, dispatch]);
  useEffect(
    () => () => {
      if (saveResult) {
        onSuccessConfirm();
        setModal(false);
      }
    },
    [saveResult, onSuccessConfirm]
  );
  useEffect(
    () => () => {
      if (playlistInfo) {
        onClearPlaylist();
        onSetTracks({ playlist: [] });
      }
    }, //eslint-disable-next-line
    [playlistInfo]
  );

  const modal = useMemo(
    () =>
      saveResult ? (
        <Modal open={isModal} clicked={() => confirmModalHandler(isModal)}>
          <div>
            <p style={{ color: "green" }}>Playlist Save Succesful!</p>
            <Button
              btnType="green"
              clicked={() => confirmModalHandler(isModal)}
            >
              Confirm
            </Button>
          </div>
        </Modal>
      ) : (
        <Modal open={isModal} clicked={() => modalHandler(isModal)}>
          <Checkout
            playlistInfo={playlistInfo ? playlistInfo.payload : null}
            cancel={() => modalHandler(isModal)}
            confirm={(payload: PlaylistPayload) => savePlaylistHandler(payload)}
          />
        </Modal>
      ),
    [
      isModal,
      saveResult,
      confirmModalHandler,
      playlistInfo,
      savePlaylistHandler,
    ]
  );

  const searchBar = useMemo(
    () => (
      <div className={classes.SearchBar}>
        <SearchBar
          clicked={(val: string, limit: number) =>
            searchHandler("track", val, limit)
          }
          limit={searchResultsLimit}
        />
      </div>
    ),
    [searchHandler, searchResultsLimit]
  );

  const searchResults = useMemo(() => <SearchResults tracklist={results} />, [
    results,
  ]);

  const playlist = useMemo(
    () => (
      <Playlist
        tracklist={tracks}
        name={playlistInfo ? playlistInfo.payload.name : null}
        clicked={() => modalHandler(isModal)}
      />
    ), //eslint-disable-next-line
    [tracks, playlistInfo]
  );

  const player = useMemo(
    () =>
      playerState && (
        <Player
          url={
            playerState[2]
              ? tracks[playerState[0]]?.preview_url
              : results[playerState[0]]?.preview_url
          }
          playerState={playerState}
        />
      ),
    [results, tracks, playerState]
  );

  return !error ? (
    !loading ? (
      <div className={classes.PlaylistMaker}>
        {modal}
        {searchBar}
        <div className={classes.Playlist}>
          <DragDropContext onDragEnd={dragEndHandler}>
            {searchResults}
            {playlist}
          </DragDropContext>
          {player}
        </div>
      </div>
    ) : (
      <Spinner />
    )
  ) : (
    <p>{error + ""}</p>
  );
};

export default errorHandler(PlaylistMaker, axios);
