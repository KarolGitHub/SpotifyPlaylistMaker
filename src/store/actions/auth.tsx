import * as actionTypes from "./actionsTypes";
import { prefixURL, client_id, redirect_uri} from "../../axios-spotify";
import { Dispatch } from "redux";
import { AxiosResponse, AxiosError } from "axios";
import axios from "../../axios-spotify";

export const authStart = () => {
  return {
    type: actionTypes.AUTH_START,
  };
};

export const authSuccess = (token: string, id: string) => {
  return {
    type: actionTypes.AUTH_SUCCESS,
    token: token,
    userId: id,
  };
};

export const authFail = (error: string) => {
  return {
    type: actionTypes.AUTH_FAIL,
    error: error,
  };
};

export const setAuthRedirectURL = () => {
  const queryParams =
    "?client_id=" +
    client_id +
    "&redirect_uri=" +
    redirect_uri +
    "&scope=" +
    "playlist-modify-public playlist-modify-private playlist-read-private playlist-read-collaborative" +
    "&response_type=token" +
    "&state=123";
  // "&show_dialog=true";
  const url = prefixURL + "authorize" + queryParams;
  return {
    type: actionTypes.SET_AUTH_REDIRECT_URL,
    url: url,
  };
};

export const auth = () => {
  return (dispatch: Dispatch) => {
    dispatch(authStart());
    const error = window.location.href.match(/error=([^&]*)/);
    if (!error) {
      const accessTokenMatch = window.location.href.match(
        /access_token=([^&]*)/
      );
      const expiresInMatch = window.location.href.match(/expires_in=([^&]*)/);
      const validateState = window.location.href.match(/state=([^&]*)/);
      if (
        accessTokenMatch &&
        expiresInMatch &&
        validateState &&
        validateState[1] === "123"
      ) {
        const accessToken = accessTokenMatch[1];
        const expiresIn = Number(expiresInMatch[1]);
        const date = new Date();
        const headers = {
          Authorization: `Bearer ${accessToken}`,
          "Content-Type": "application/json",
        };
        axios
          .get("/me", { headers: headers })
          .then((response) => JSON.parse(JSON.stringify(response)))
          .then((response: AxiosResponse) => {
            localStorage.setItem("userId", response.data.id);
            localStorage.setItem(
              "expirationDate",
              date.setSeconds(date.getSeconds() + expiresIn).toString()
            );
            localStorage.setItem("token", accessToken);
          })
          .then(() => window.close())
          .catch((error: AxiosError) => {
            if (error.response?.data) {
              dispatch(authFail(error.response.data.error.message));
            } else if (error.message) {
              dispatch(authFail(error.message));
            } else {
              dispatch(authFail("Unexpected Error!"));
            }
          });
      } else {
        dispatch(authFail("Authentication Error"));
      }
    } else {
      dispatch(authFail(error[1]));
      window.close();
    }
  };
};

export const checkAuthTimeout = (expirationTime: number) => {
  return (dispatch: Dispatch) => {
    setTimeout(() => {
      dispatch(logout());
    }, (expirationTime - 5) * 1000);
  };
};

export const logout = () => {
  localStorage.removeItem("token");
  localStorage.removeItem("userId");
  localStorage.removeItem("expirationDate");
  return {
    type: actionTypes.AUTH_LOGOUT,
  };
};

export const authCheckState = () => {
  return (dispatch: Dispatch) => {
    window.removeEventListener("storage", () => { });
    const token = localStorage.getItem("token");
    const userId = localStorage.getItem("userId");
    if (!token || !userId) {
      dispatch(logout());
    } else {
      let expirationDate = new Date(
        Number(localStorage.getItem("expirationDate"))
      );
      if (expirationDate <= new Date()) {
        dispatch(logout());
      } else {
        dispatch(authSuccess(token, userId));
        checkAuthTimeout(
          (expirationDate.getTime() - Number(new Date().getTime())) / 1000
        );
      }
    }
  };
};
