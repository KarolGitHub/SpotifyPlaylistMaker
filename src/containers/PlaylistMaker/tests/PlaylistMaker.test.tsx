import React from "react";
import { shallow, mount } from "enzyme";
import configureMockStore from 'redux-mock-store';
import * as actionTypes from '../../../store/actions/actionsTypes';
import * as actions from '../../../store/actions/index';
import { initialState as playlistMakerState } from '../../../store/reducers/playlistMaker';
import { initialState as playlistsState } from '../../../store/reducers/playlists';
import { initialState as playerState } from '../../../store/reducers/player';
import { initialState as authState } from '../../../store/reducers/auth';
import PlaylistMaker from "../PlaylistMaker";
import Playlist from "../Playlist/Playlist";
import { Provider } from "react-redux";
import Button from "../../../components/UI/Button/Button";
import Input from "../../../components/UI/Input/Input";
import { act } from "react-dom/test-utils";

const middlewares: any = [];
const mockStore = configureMockStore(middlewares);

describe("<PlaylistMaker />", () => {
  describe("Unit tests", () => {
    let wrapper: any;
    beforeEach(() => {
      wrapper = shallow(<PlaylistMaker />);
    });

    it('render without crashing', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it("should render <Playlist />", () => {
      expect(wrapper.find(Playlist)).toEqual({});
    });

  });
  describe("Integration tests", () => {
    let store: any;
    beforeEach(() => {
      store = mockStore({
        playlistMaker: playlistMakerState,
        auth: authState,
        player: playerState,
        playlists: playlistsState
      });
    });

    it("should dispatch an action on button click", () => {
      act(() => {
        const wrapper = mount(
          <Provider store={store}>
            <PlaylistMaker />
          </Provider>
        );
        store.dispatch(actions.authSuccess('token', 'id'));
        const asyncCheck = setImmediate(() => {
          wrapper.update();
          expect(store.getState()).toContainEqual({
            token: 'token',
            userId: 'id',
            error: null,
            loading: false,
            authRedirectPath: 'https://accounts.spotify.com/authorize?client_id=4ab35e6f4e00492f9af5a2358e409f9f&redirect_uri=http://localhost:3000/login&scope=playlist-modify-public playlist-modify-private playlist-read-private playlist-read-collaborative&response_type=token&state=123',
          });
        });
        wrapper.find(Input).at(4).props().changed({ target: { value: 'text' } });
        wrapper.find(Button).at(2).simulate('click');
        const asyncCheck2 = setImmediate(() => {
          wrapper.update();
          expect(store.getActions()).toContainEqual([{
            type: actionTypes.SEARCH_TRACKS,
            token: 'token',
            userId: 'id',
            term: 'text',
            limit: 10,
          }]);
        });
        global.clearImmediate(asyncCheck);
        global.clearImmediate(asyncCheck2);
        wrapper.unmount();
      });
    });
  });
});
