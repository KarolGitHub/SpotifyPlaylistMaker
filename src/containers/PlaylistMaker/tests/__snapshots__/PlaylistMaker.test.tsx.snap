// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistMaker /> Integration tests should render with given state from Redux store 1`] = `ReactWrapper {}`;

exports[`<PlaylistMaker /> Unit tests render without crashing 1`] = `ShallowWrapper {}`;

exports[`<PlaylistMaker /> Unit tests should render with given state from Redux store 1`] = `undefined`;

exports[`<PlaylistMaker /> Unit tests snapshot match 1`] = `ShallowWrapper {}`;
